atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/InvalidationRunner.java

```diff
@@ -0,0 +1,192 @@
+
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Optional;
+
+import org.apache.commons.dbutils.QueryRunner;
+import org.immutables.value.Value;
+
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.keyvalue.dbkvs.timestamp.BoundStoreUtils;
+import com.palantir.atlasdb.keyvalue.dbkvs.timestamp.ConnectionDbTypes;
+import com.palantir.atlasdb.keyvalue.dbkvs.timestamp.InDbTimestampBoundStoreInitializer;
+import com.palantir.atlasdb.spi.AtlasDbFactory;
+import com.palantir.common.base.Throwables;
+import com.palantir.exception.PalantirSqlException;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import com.palantir.nexus.db.DBType;
+import com.palantir.nexus.db.pool.ConnectionManager;
+import com.palantir.nexus.db.pool.RetriableTransactions;
+
+public class InvalidationRunner {
+    private static final String LAST_ALLOCATED = "last_allocated";
+    private static final String LEGACY_LAST_ALLOCATED = "legacy_last_allocated";
+
+    private final ConnectionManager connManager;
+    private final InDbTimestampBoundStoreInitializer helper;
+
+    public InvalidationRunner(ConnectionManager connManager) {
+        this.connManager = connManager;
+        this.helper = new InDbTimestampBoundStoreInitializer(connManager);
+    }
+
+    public void createTableIfDoesNotExist() {
+        helper.createTableIfDoesNotExist(prefixedTimestampTableName());
+    }
+
+    public long ensureInDbStoreIsPoisonedAndGetLastAllocatedTimestamp() {
+        RetriableTransactions.TransactionResult<Long> result = RetriableTransactions.run(connManager,
+                connection -> {
+                    Limits limits = getLimits(connection);
+                    TableStatus tableStatus = checkTableStatus(limits);
+
+                    if (tableStatus == TableStatus.POISONED) {
+                        return limits.legacyUpperLimit().get().value();
+                    }
+                    return poisonStoreAndGetLastAllocatedTimestamp(connection, limits, tableStatus);
+                });
+
+        switch (result.getStatus()) {
+            case SUCCESSFUL:
+                return result.getResultValue();
+            case UNKNOWN:
+            case FAILED:
+                Throwable error = result.getError();
+                if (error instanceof SQLException) {
+                    throw PalantirSqlException.create((SQLException) error);
+                }
+                throw Throwables.rewrapAndThrowUncheckedException(error);
+            default:
+                throw new SafeIllegalStateException("Unrecognized transaction status.",
+                        SafeArg.of("status", result.getStatus()));
+        }
+    }
+
+    public Long poisonStoreAndGetLastAllocatedTimestamp(Connection connection, Limits limits,
```

Comment:
nit: does this one need to be public? might have missed something but it feels like a private thing!